<a href="https://colab.research.google.com/github/carlosfmorenog/CM4125/blob/main/CM4125_T6/CM4125_Lec6.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# CM4125 Topic 6: Data Manipulation (Part 1)

## Lecture objectives

1) Load a set of datasets containing different characteristics

2) Understand the best methods to manipulate it prior to analysis

3) Apply more elaborated pre-processing techniques to clean data and fill in missing values

## Data by Columns

In [ ]:
# Importing a dataset that I have in dropbox related to Time magazine MoTY
import pandas as pd
time = pd.read_csv('https://www.dropbox.com/scl/fi/muy48qj9p2hb1h5u9wk4x/times.csv?rlkey=ucnespmybj64px1woyfx32arv&raw=1')
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  
0             NaN  First Solo Transatlantic Flight  
1       Economics            Chrysler/Dodge Merger  
2       Diplomacy                       Young Plan  
3      Revolution                       Salt March  
4        Politics                              NaN  
..            ...                              ...  
94       Politics            Presidential Election  
95     Technology      Richest Person in the World  
96       Politics                   War in Ukraine  
97  Entertainment     Cultural and Economic Impact  
98       Politics            Presidential Election  

[99 rows x 9 columns]

This dataset contains the list of people named Man/Person of the Year by Times magazine

Last week, we added a new column to a dataframe (the "watched" column at the end of the Netflix dataset)

Today, we will learn other ways in which we can manipulate, add or relate data from columns

### Mapping Columns

Sometimes we want to convert a column of categorical data from strings (text) to integers (numbers)

We can define a `dictionary` which maps the strings (as keys) to integers (as values) using the `map` method

Another option is to define a `set` of the unique categories as follows.

In [ ]:
categories = set(time['Category'])
categories

{'Diplomacy',
 'Economics',
 'Entertainment',
 'Environment',
 'Media',
 'Philanthropy',
 'Politics',
 'Religion',
 'Revolution',
 'Science',
 'Space',
 'Technology',
 'War',
 nan}

Then, we can then assign unique numbers to this `set` as a `dictionary` (defined in Python using the `{}`)

In [ ]:
mapping = {category : index for index, category in enumerate(categories)}
mapping

{'Revolution': 0,
 'Technology': 1,
 'Politics': 2,
 'War': 3,
 'Media': 4,
 'Environment': 5,
 'Economics': 6,
 'Diplomacy': 7,
 nan: 8,
 'Space': 9,
 'Science': 10,
 'Philanthropy': 11,
 'Entertainment': 12,
 'Religion': 13}

We now have a `map` which allows us to convert from a category string to an `int`. For example:

In [ ]:
mapping['Diplomacy']

7

In [ ]:
mapping['Media']

4

We can use the `map` method on the chosen column to do the conversion

The `map` method takes our `dictionary` as an argument and returns a new column with the mapping applied

In [ ]:
time['Category'].map(mapping)

0      8
1      6
2      7
3      0
4      2
      ..
94     2
95     1
96     2
97    12
98     2
Name: Category, Length: 99, dtype: int64

If we wanted to overwrite the `Category` column with our new mapped values, we could do:

    time['Category'] = time['Category'].map(mapping)

However, it may be best to keep both. So we can instead assign it a different column name

In [ ]:
time['Category Ordinal'] = time['Category'].map(mapping)
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  Category Ordinal  
0             NaN  First Solo Transatlantic Flight                 8  
1       Economics            Chrysler/Dodge Merger                 6  
2       Diplomacy                       Young Plan                 7  
3      Revolution                       Salt March                 0  
4        Politics                              NaN                 2  
..            ...                              ...               ...  
94       Politics            Presidential Election                 2  
95     Technology      Richest Person in the World                 1  
96       Politics                   War in Ukraine                 2  
97  Entertainment     Cultural and Economic Impact                12  
98       Politics            Presidential Election                 2  

[99 rows x 10 columns]

### Manipulating Numerical Columns

We can perform arithmetic between columns as if they were simply numbers

We will take the column `Death Year`, subtract the column `Birth Year`, and storing the result in a (new) column called `Lifespan`

In [ ]:
time['Lifespan'] = time['Death Year'] - time['Birth Year']
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  Category Ordinal  Lifespan  
0             NaN  First Solo Transatlantic Flight                 8      72.0  
1       Economics            Chrysler/Dodge Merger                 6      65.0  
2       Diplomacy                       Young Plan                 7      88.0  
3      Revolution                       Salt March                 0      79.0  
4        Politics                              NaN                 2      62.0  
..            ...                              ...               ...       ...  
94       Politics            Presidential Election                 2       NaN  
95     Technology      Richest Person in the World                 1       NaN  
96       Politics                   War in Ukraine                 2       NaN  
97  Entertainment     Cultural and Economic Impact                12       NaN  
98       Politics            Presidential Election                 2       NaN  

[99 rows x 11 columns]

Notice that the method acknowledged that some columns were `NaN` (since the person is still alive) and thus put a `NaN` as a result in the lifespan column.

Sometimes you want to do "more difficult" numerical conversions

In those cases, it is more convenient to define a **function** and then use the `apply` method

In this example, we will define a function to convert years to decades to be applied in the different columns that contain years

In [ ]:
def to_decade(value):
    return 10 * (value // 10) # // is a floor division

For instance, 1965 // 10 = 196, and then * 10 is equal to 1960

First, we should test our function on some examples to ensure the function is correct.

In [ ]:
to_decade(1971)

1970

Now, we can use `apply` to transform the entire column `Birth Year` into the corresponding decade

In [ ]:
time['Birth Decade'] = time['Birth Year'].apply(to_decade)
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  Category Ordinal  \
0             NaN  First Solo Transatlantic Flight                 8   
1       Economics            Chrysler/Dodge Merger                 6   
2       Diplomacy                       Young Plan                 7   
3      Revolution                       Salt March                 0   
4        Politics                              NaN                 2   
..            ...                              ...               ...   
94       Politics            Presidential Election                 2   
95     Technology      Richest Person in the World                 1   
96       Politics                   War in Ukraine                 2   
97  Entertainment     Cultural and Economic Impact                12   
98       Politics            Presidential Election                 2   

    Lifespan  Birth Decade  
0       72.0        1900.0  
1       65.0        1870.0  
2       88.0        1870.0  
3       79.0        1860.0  
4       62.0        1880.0  
..       ...           ...  
94       NaN           NaN  
95       NaN        1970.0  
96       NaN        1970.0  
97       NaN        1980.0  
98       NaN        1940.0  

[99 rows x 12 columns]

### Manipulating String Columns

Even if you use tableau/PowerBi, you need to use this some times (example of inserting calculated fields in tableau)

`apply` can also be used on strings

For example, we can apply the `len` function to `Name` to get the number of characters in each name.

In [ ]:
time['Name Length'] = time['Name'].apply(len)
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  Category Ordinal  \
0             NaN  First Solo Transatlantic Flight                 8   
1       Economics            Chrysler/Dodge Merger                 6   
2       Diplomacy                       Young Plan                 7   
3      Revolution                       Salt March                 0   
4        Politics                              NaN                 2   
..            ...                              ...               ...   
94       Politics            Presidential Election                 2   
95     Technology      Richest Person in the World                 1   
96       Politics                   War in Ukraine                 2   
97  Entertainment     Cultural and Economic Impact                12   
98       Politics            Presidential Election                 2   

    Lifespan  Birth Decade  Name Length  
0       72.0        1900.0           17  
1       65.0        1870.0           15  
2       88.0        1870.0           13  
3       79.0        1860.0           14  
4       62.0        1880.0           12  
..       ...           ...          ...  
94       NaN           NaN           25  
95       NaN        1970.0            9  
96       NaN        1970.0           18  
97       NaN        1980.0           12  
98       NaN        1940.0           12  

[99 rows x 13 columns]

Another example, we can create a function to get the initials from a name.

In [ ]:
# First we define the function
def to_initials(name):
    import numpy as np # we need this for Python to know what NaN is!
    if name == np.nan:
        return np.nan
    else:
        initials = ""
        for word in name.split(' '):
            first_letter = word[0]
            initials += first_letter
        return initials
    # Then, we try it
to_initials("Barack Obama")

'BO'

Applying this gives us a name column with the person's initials.

In [ ]:
time['Initials'] = time['Name'].apply(to_initials)
time

Year               Honor                       Name        Country  \
0   1927     Man of the Year          Charles Lindbergh  United States   
1   1928     Man of the Year            Walter Chrysler  United States   
2   1929     Man of the Year              Owen D. Young  United States   
3   1930     Man of the Year             Mahatma Gandhi          India   
4   1931     Man of the Year               Pierre Laval         France   
..   ...                 ...                        ...            ...   
94  2020  Person of the Year  Joe Biden & Kamala Harris  United States   
95  2021  Person of the Year                  Elon Musk   South Africa   
96  2022  Person of the Year         Volodymyr Zelensky        Ukraine   
97  2023  Person of the Year               Taylor Swift  United States   
98  2024  Person of the Year               Donald Trump  United States   

    Birth Year  Death Year                                              Title  \
0       1902.0      1974.0                                  US Air Mail Pilot   
1       1875.0      1940.0                                Founder of Chrysler   
2       1874.0      1962.0  Member of the German Reparations International...   
3       1869.0      1948.0                                                NaN   
4       1883.0      1945.0                           Prime Minister of France   
..         ...         ...                                                ...   
94         NaN         NaN            President and Vice President of the USA   
95      1971.0         NaN                                       CEO of Tesla   
96      1978.0         NaN                               President of Ukraine   
97      1989.0         NaN                                  Singer-songwriter   
98      1946.0         NaN                     President of the United States   

         Category                          Context  Category Ordinal  \
0             NaN  First Solo Transatlantic Flight                 8   
1       Economics            Chrysler/Dodge Merger                 6   
2       Diplomacy                       Young Plan                 7   
3      Revolution                       Salt March                 0   
4        Politics                              NaN                 2   
..            ...                              ...               ...   
94       Politics            Presidential Election                 2   
95     Technology      Richest Person in the World                 1   
96       Politics                   War in Ukraine                 2   
97  Entertainment     Cultural and Economic Impact                12   
98       Politics            Presidential Election                 2   

    Lifespan  Birth Decade  Name Length Initials  
0       72.0        1900.0           17       CL  
1       65.0        1870.0           15       WC  
2       88.0        1870.0           13      ODY  
3       79.0        1860.0           14       MG  
4       62.0        1880.0           12       PL  
..       ...           ...          ...      ...  
94       NaN           NaN           25    JB&KH  
95       NaN        1970.0            9       EM  
96       NaN        1970.0           18       VZ  
97       NaN        1980.0           12       TS  
98       NaN        1940.0           12       DT  

[99 rows x 14 columns]

### Removing Columns (Unwanted Variables)

There are two main ways to remove them, either we specify which columns we want to **keep** or we specify which we want to **remove**

Here is the version where we specify, by name, which columns to **keep**.

In [ ]:
time[['Year', 'Honor', 'Name']]

Year               Honor                       Name
0   1927     Man of the Year          Charles Lindbergh
1   1928     Man of the Year            Walter Chrysler
2   1929     Man of the Year              Owen D. Young
3   1930     Man of the Year             Mahatma Gandhi
4   1931     Man of the Year               Pierre Laval
..   ...                 ...                        ...
94  2020  Person of the Year  Joe Biden & Kamala Harris
95  2021  Person of the Year                  Elon Musk
96  2022  Person of the Year         Volodymyr Zelensky
97  2023  Person of the Year               Taylor Swift
98  2024  Person of the Year               Donald Trump

[99 rows x 3 columns]

Here is the version where we specify, by name, which columns to **remove**

In [ ]:
time.drop(columns=['Year', 'Honor', 'Name'])

Country  Birth Year  Death Year  \
0   United States      1902.0      1974.0   
1   United States      1875.0      1940.0   
2   United States      1874.0      1962.0   
3           India      1869.0      1948.0   
4          France      1883.0      1945.0   
..            ...         ...         ...   
94  United States         NaN         NaN   
95   South Africa      1971.0         NaN   
96        Ukraine      1978.0         NaN   
97  United States      1989.0         NaN   
98  United States      1946.0         NaN   

                                                Title       Category  \
0                                   US Air Mail Pilot            NaN   
1                                 Founder of Chrysler      Economics   
2   Member of the German Reparations International...      Diplomacy   
3                                                 NaN     Revolution   
4                            Prime Minister of France       Politics   
..                                                ...            ...   
94            President and Vice President of the USA       Politics   
95                                       CEO of Tesla     Technology   
96                               President of Ukraine       Politics   
97                                  Singer-songwriter  Entertainment   
98                     President of the United States       Politics   

                            Context  Category Ordinal  Lifespan  Birth Decade  \
0   First Solo Transatlantic Flight                 8      72.0        1900.0   
1             Chrysler/Dodge Merger                 6      65.0        1870.0   
2                        Young Plan                 7      88.0        1870.0   
3                        Salt March                 0      79.0        1860.0   
4                               NaN                 2      62.0        1880.0   
..                              ...               ...       ...           ...   
94            Presidential Election                 2       NaN           NaN   
95      Richest Person in the World                 1       NaN        1970.0   
96                   War in Ukraine                 2       NaN        1970.0   
97     Cultural and Economic Impact                12       NaN        1980.0   
98            Presidential Election                 2       NaN        1940.0   

    Name Length Initials  
0            17       CL  
1            15       WC  
2            13      ODY  
3            14       MG  
4            12       PL  
..          ...      ...  
94           25    JB&KH  
95            9       EM  
96           18       VZ  
97           12       TS  
98           12       DT  

[99 rows x 11 columns]

To apply the changes, assign the result back to the same `time` variable

In [ ]:
time = time.drop(columns=['Year', 'Honor', 'Name'])
time

Country  Birth Year  Death Year  \
0   United States      1902.0      1974.0   
1   United States      1875.0      1940.0   
2   United States      1874.0      1962.0   
3           India      1869.0      1948.0   
4          France      1883.0      1945.0   
..            ...         ...         ...   
94  United States         NaN         NaN   
95   South Africa      1971.0         NaN   
96        Ukraine      1978.0         NaN   
97  United States      1989.0         NaN   
98  United States      1946.0         NaN   

                                                Title       Category  \
0                                   US Air Mail Pilot            NaN   
1                                 Founder of Chrysler      Economics   
2   Member of the German Reparations International...      Diplomacy   
3                                                 NaN     Revolution   
4                            Prime Minister of France       Politics   
..                                                ...            ...   
94            President and Vice President of the USA       Politics   
95                                       CEO of Tesla     Technology   
96                               President of Ukraine       Politics   
97                                  Singer-songwriter  Entertainment   
98                     President of the United States       Politics   

                            Context  Category Ordinal  Lifespan  Birth Decade  \
0   First Solo Transatlantic Flight                 8      72.0        1900.0   
1             Chrysler/Dodge Merger                 6      65.0        1870.0   
2                        Young Plan                 7      88.0        1870.0   
3                        Salt March                 0      79.0        1860.0   
4                               NaN                 2      62.0        1880.0   
..                              ...               ...       ...           ...   
94            Presidential Election                 2       NaN           NaN   
95      Richest Person in the World                 1       NaN        1970.0   
96                   War in Ukraine                 2       NaN        1970.0   
97     Cultural and Economic Impact                12       NaN        1980.0   
98            Presidential Election                 2       NaN        1940.0   

    Name Length Initials  
0            17       CL  
1            15       WC  
2            13      ODY  
3            14       MG  
4            12       PL  
..          ...      ...  
94           25    JB&KH  
95            9       EM  
96           18       VZ  
97           12       TS  
98           12       DT  

[99 rows x 11 columns]

Some functions/methods also have a parameter called `inplace` which, if set = True, apply the change to the original variable

In [ ]:
time.drop(columns=['Lifespan'], inplace=True)
time

Country  Birth Year  Death Year  \
0   United States      1902.0      1974.0   
1   United States      1875.0      1940.0   
2   United States      1874.0      1962.0   
3           India      1869.0      1948.0   
4          France      1883.0      1945.0   
..            ...         ...         ...   
94  United States         NaN         NaN   
95   South Africa      1971.0         NaN   
96        Ukraine      1978.0         NaN   
97  United States      1989.0         NaN   
98  United States      1946.0         NaN   

                                                Title       Category  \
0                                   US Air Mail Pilot            NaN   
1                                 Founder of Chrysler      Economics   
2   Member of the German Reparations International...      Diplomacy   
3                                                 NaN     Revolution   
4                            Prime Minister of France       Politics   
..                                                ...            ...   
94            President and Vice President of the USA       Politics   
95                                       CEO of Tesla     Technology   
96                               President of Ukraine       Politics   
97                                  Singer-songwriter  Entertainment   
98                     President of the United States       Politics   

                            Context  Category Ordinal  Birth Decade  \
0   First Solo Transatlantic Flight                 8        1900.0   
1             Chrysler/Dodge Merger                 6        1870.0   
2                        Young Plan                 7        1870.0   
3                        Salt March                 0        1860.0   
4                               NaN                 2        1880.0   
..                              ...               ...           ...   
94            Presidential Election                 2           NaN   
95      Richest Person in the World                 1        1970.0   
96                   War in Ukraine                 2        1970.0   
97     Cultural and Economic Impact                12        1980.0   
98            Presidential Election                 2        1940.0   

    Name Length Initials  
0            17       CL  
1            15       WC  
2            13      ODY  
3            14       MG  
4            12       PL  
..          ...      ...  
94           25    JB&KH  
95            9       EM  
96           18       VZ  
97           12       TS  
98           12       DT  

[99 rows x 10 columns]

### Renaming Columns

You can do this using a `dictionary`

In [ ]:
time = time.rename(columns={'Birth Year' : 'Born',
                        'Death Year' : 'Died'})
time

Country    Born    Died  \
0   United States  1902.0  1974.0   
1   United States  1875.0  1940.0   
2   United States  1874.0  1962.0   
3           India  1869.0  1948.0   
4          France  1883.0  1945.0   
..            ...     ...     ...   
94  United States     NaN     NaN   
95   South Africa  1971.0     NaN   
96        Ukraine  1978.0     NaN   
97  United States  1989.0     NaN   
98  United States  1946.0     NaN   

                                                Title       Category  \
0                                   US Air Mail Pilot            NaN   
1                                 Founder of Chrysler      Economics   
2   Member of the German Reparations International...      Diplomacy   
3                                                 NaN     Revolution   
4                            Prime Minister of France       Politics   
..                                                ...            ...   
94            President and Vice President of the USA       Politics   
95                                       CEO of Tesla     Technology   
96                               President of Ukraine       Politics   
97                                  Singer-songwriter  Entertainment   
98                     President of the United States       Politics   

                            Context  Category Ordinal  Birth Decade  \
0   First Solo Transatlantic Flight                 8        1900.0   
1             Chrysler/Dodge Merger                 6        1870.0   
2                        Young Plan                 7        1870.0   
3                        Salt March                 0        1860.0   
4                               NaN                 2        1880.0   
..                              ...               ...           ...   
94            Presidential Election                 2           NaN   
95      Richest Person in the World                 1        1970.0   
96                   War in Ukraine                 2        1970.0   
97     Cultural and Economic Impact                12        1980.0   
98            Presidential Election                 2        1940.0   

    Name Length Initials  
0            17       CL  
1            15       WC  
2            13      ODY  
3            14       MG  
4            12       PL  
..          ...      ...  
94           25    JB&KH  
95            9       EM  
96           18       VZ  
97           12       TS  
98           12       DT  

[99 rows x 10 columns]

### How to create new columns in Tableau?

## Missing Data

### Fundamentals of missing data

In the previous example, you saw that some entries had `NaN` to fill in for **missing data**

Missing data is one of the most recurrent problems in data analysis

In fact, there are a lot of studies and methods on how to handle it!

We will work with a small dataset called [missing.csv](https://www.dropbox.com/s/5h4k4rszebd6p0n/missing.csv?raw=1) (it's the same as data.csv from last week, but now with missing values)

In [ ]:
df = pd.read_csv('https://www.dropbox.com/scl/fi/4n7in0fnv7qtcrumfk43w/missing.csv?rlkey=g1ql9y050qbpfiz0zjd39628t&raw=1')
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

The first thing to notice is that the second column is being shown with decimals, even though they were integers in the original file

Moreover, the blank entry is displayed as `NaN`

`NaN` stands for *Not A Number* as is part of the floating point specification, typically used when a calculation has no valid numerical result

In fact, Python considers `NaN` as a floating point (or `float64`)

In [ ]:
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

In [ ]:
df.at[3, 'Age']

np.float64(nan)

As a result, the entire column has been turned into `float64`, instead of `int64`

 This is why it was being shown as decimals!

In [ ]:
df['Age']

0    21.0
1    29.0
2    28.0
3     NaN
4    35.0
5    21.0
Name: Age, dtype: float64

### Incorrectly Imported Missing Data

The missing value in the `Age` column was generated because the cell was blank

By default, `Pandas` will treat certain cell values, such as a *blank*, `NULL`, `NaN`, and `n/a` as missing values

The `Height` column contains some entries which we recognise as missing data (i.e. `Unknown`) but that Pandas did not

In [ ]:
df.at[1, 'Height']

'Unknown'

In fact, the entire column has been imported as strings, not numbers!

In [ ]:
df

Name   Age   Height
0       Nick  21.0     1.85
1      Chris  29.0  Unknown
2        Tim  28.0     1.75
3        Ron   NaN     1.81
4     Monica  35.0  Unknown
5  Cassandra  21.0     1.66

In [ ]:
print(df.at[1, 'Height'])
print(type(df.at[1, 'Height']))

Unknown
<class 'str'>


This in turn made `Pandas` believe that the whole column has strings!

In [ ]:
print(df.at[0, 'Height'])
print(type(df.at[0, 'Height']))

1.85
<class 'str'>


What happens if we multiply this column by `100`?

In [ ]:
df['Height times 100'] = df['Height'] * 100
df

Name   Age   Height                                   Height times 100
0       Nick  21.0     1.85  1.851.851.851.851.851.851.851.851.851.851.851....
1      Chris  29.0  Unknown  UnknownUnknownUnknownUnknownUnknownUnknownUnkn...
2        Tim  28.0     1.75  1.751.751.751.751.751.751.751.751.751.751.751....
3        Ron   NaN     1.81  1.811.811.811.811.811.811.811.811.811.811.811....
4     Monica  35.0  Unknown  UnknownUnknownUnknownUnknownUnknownUnknownUnkn...
5  Cassandra  21.0     1.66  1.661.661.661.661.661.661.661.661.661.661.661....

We need to tell Pandas in advance to treat `Unknown` not as a string, but as a missing value

Lets re-import the data set passing a list of values which represent missing values as the `na_values` parameter

In [ ]:
df = pd.read_csv('https://www.dropbox.com/scl/fi/4n7in0fnv7qtcrumfk43w/missing.csv?rlkey=g1ql9y050qbpfiz0zjd39628t&raw=1',
                 na_values=['Unknown'])
df

Name   Age  Height
0       Nick  21.0    1.85
1      Chris  29.0     NaN
2        Tim  28.0    1.75
3        Ron   NaN    1.81
4     Monica  35.0     NaN
5  Cassandra  21.0    1.66

Now the column is `float64`

In [ ]:
df['Height']

0    1.85
1     NaN
2    1.75
3    1.81
4     NaN
5    1.66
Name: Height, dtype: float64

Multiplying `Height` now works as expected.

In [ ]:
df['Height times 100'] = df['Height'] * 100
df

Name   Age  Height  Height times 100
0       Nick  21.0    1.85             185.0
1      Chris  29.0     NaN               NaN
2        Tim  28.0    1.75             175.0
3        Ron   NaN    1.81             181.0
4     Monica  35.0     NaN               NaN
5  Cassandra  21.0    1.66             166.0

In [ ]:
# Another example
df['Test'] = df['Age'] + df['Height']
df

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
3        Ron   NaN    1.81             181.0    NaN
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

### Methods to address missing data

As mentioned before, there are different ways to deal with missing data

The method you should use depends on your analysis, the application domain and what you want to achieve

Here are some common methods:

    1. Leave the missing values as NaN
    2. Lookup the values from another data source
    3. Delete the rows / observations
    4. Compute reasonable guesses about the values

In **option 1**, it is often reasonable to just leave the data as missing data

For example, if we wanted to calculate average `Height`, it may not matter that we are missing `Age` for one row

As for **option 2**, if we have some other ways of going back and getting correct data, we could just update the values with `df.at[3, 'Age'] = ...`

Let's assume we are going for **option 3** and want to remove any rows where we do no have `Age`

If we know the row we want to delete (`3`), we can use the `drop` method, only this time we specify rows, not columns. As in:

    df.drop(3)

However, often there are many missing values, and we don't want to check on the index of each

A good way to remove rows with missing data is to specify which rows to **keep**

The `isnull` method will filter for rows which contains `NaN`.

In [ ]:
df[df['Age'].isnull()]

Name  Age  Height  Height times 100  Test
3  Ron  NaN    1.81             181.0   NaN

Of course, this is the *opposite* of specifying which to keep

The **notnull** method will show us which rows do not contain `NaN` in the specified column.

In [ ]:
df[df['Age'].notnull()]

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

Notice that the filter is only checking the column we specified `Age`, not the other column with `NaN` values

We can apply the filter by assigning the result back to `df`

In [ ]:
df = df[df['Age'].notnull()]
df

Name   Age  Height  Height times 100   Test
0       Nick  21.0    1.85             185.0  22.85
1      Chris  29.0     NaN               NaN    NaN
2        Tim  28.0    1.75             175.0  29.75
4     Monica  35.0     NaN               NaN    NaN
5  Cassandra  21.0    1.66             166.0  22.66

Since the `Age` column no longer contains any non-integers, it is safe to convert the column to `int64` type

We can do this with the `astype` method

In [ ]:
df = df.astype({'Age': 'int64'})
df

Name  Age  Height  Height times 100   Test
0       Nick   21    1.85             185.0  22.85
1      Chris   29     NaN               NaN    NaN
2        Tim   28    1.75             175.0  29.75
4     Monica   35     NaN               NaN    NaN
5  Cassandra   21    1.66             166.0  22.66

### Updating Indices

After these operations, the current data frame has the index 0, 1, 2, 4, 5

We could leave it like this, being careful to select the correct rows with `loc`. However, we can also reset the index with `reset_index`

Note that we need to specify `drop=True`, without this we will end up with an extra column with the old index

In [ ]:
df = df.reset_index(drop=True)
df

Name  Age  Height  Height times 100   Test
0       Nick   21    1.85             185.0  22.85
1      Chris   29     NaN               NaN    NaN
2        Tim   28    1.75             175.0  29.75
3     Monica   35     NaN               NaN    NaN
4  Cassandra   21    1.66             166.0  22.66

Alternatively, we can set the index to a particular column if we don't want default indexing

In [ ]:
df = df.set_index('Name')
df

Age  Height  Height times 100   Test
Name                                           
Nick        21    1.85             185.0  22.85
Chris       29     NaN               NaN    NaN
Tim         28    1.75             175.0  29.75
Monica      35     NaN               NaN    NaN
Cassandra   21    1.66             166.0  22.66

Say we want to go with **option 4** and replace missing `Height` with a computed value

In this example we simply calculate the mean height of other people in the data set (to the nearest cm)

**Note:** The `mean` method just ignores the missing data and gives us the average of the non-missing data, which is what we want

In [ ]:
mean_height = df['Height'].mean()
mean_height

np.float64(1.7533333333333332)

In [ ]:
mean_height = round(mean_height, 2)
mean_height

np.float64(1.75)

Of course there are many more sophisticated ways to predict a replacement value!

For instance, could fit a regression with `Age`

Or if we had categories such as `Gender` we could group by categories and do predictions for each

In this example we will just use `1.75` for every missing value.

In [ ]:
df = df.fillna({'Height' : mean_height, 'Height times 100':175})
df

Age  Height  Height times 100   Test
Name                                           
Nick        21    1.85             185.0  22.85
Chris       29    1.75             175.0    NaN
Tim         28    1.75             175.0  29.75
Monica      35    1.75             175.0    NaN
Cassandra   21    1.66             166.0  22.66

## Lab